File: SundeeFundee/Sources/SundeeFundeeKit/DomainLayer/Intelligence/WeeklyLoadAnalyzer.swift
import Foundation

// MARK: - WeeklyLoadAnalyzer

/// Analyzes completed workout history to surface volume, intensity,
/// frequency, and muscle-group balance trends over rolling weeks.
///
/// Pure domain logic — no framework dependencies.
public enum WeeklyLoadAnalyzer {

    // MARK: - Types

    /// Summary of a single training week.
    public struct WeeklySummary: Sendable, Equatable {
        /// The Monday that starts this ISO week.
        public let weekStartDate: Date

        /// Total number of workouts completed.
        public let workoutCount: Int

        /// Total training duration in minutes.
        public let totalMinutes: Int

        /// Distinct muscle groups trained (from exercise names).
        public let muscleGroupsHit: Set<String>

        /// Exercises performed this week.
        public let exerciseNames: [String]
    }

    /// A detected trend across multiple weeks.
    public struct LoadTrend: Sendable, Equatable {
        public let type: TrendType
        public let message: String
        public let severity: TrendSeverity
    }

    public enum TrendType: String, Sendable, Equatable {
        /// Workout frequency is declining.
        case frequencyDrop
        /// Workout frequency is increasing nicely.
        case frequencyRise
        /// Muscle group imbalance detected.
        case muscleImbalance
        /// Possible overreaching (too much volume).
        case overreaching
        /// Undertraining (very low volume).
        case undertraining
        /// Consistent training pattern.
        case consistent
        /// Push/pull ratio imbalance within a week.
        case pushPullImbalance
        /// Repeating pattern of missed/dropped weeks.
        case missedSessionPattern
    }

    public enum TrendSeverity: String, Sendable, Equatable {
        case info
        case warning
        case positive
    }

    // MARK: - Analysis

    /// Builds weekly summaries from completed workout records.
    ///
    /// - Parameters:
    ///   - workouts: Completed workout records.
    ///   - weekCount: Number of recent weeks to analyze (default 4).
    /// - Returns: Weekly summaries sorted oldest-first.
    public static func weeklySummaries(
        from workouts: [CompletedWorkoutRecord],
        weekCount: Int = 4
    ) -> [WeeklySummary] {
        let calendar = Calendar.current

        // Group workouts by ISO week start (Monday)
        var weekBuckets: [Date: [CompletedWorkoutRecord]] = [:]

        for workout in workouts {
            guard let weekStart = calendar.dateInterval(of: .weekOfYear, for: workout.date)?.start else {
                continue
            }
            weekBuckets[weekStart, default: []].append(workout)
        }

        // Build summaries for the most recent N weeks
        let sortedWeeks = weekBuckets.keys.sorted()
        let recentWeeks = Array(sortedWeeks.suffix(weekCount))

        return recentWeeks.map { weekStart in
            let records = weekBuckets[weekStart] ?? []
            let allExercises = records.flatMap { $0.exerciseNames }
            let muscleGroups = Set(allExercises.compactMap { classifyMuscleGroup($0) })
            let totalMinutes = records.compactMap { $0.duration }.reduce(0, +)

            return WeeklySummary(
                weekStartDate: weekStart,
                workoutCount: records.count,
                totalMinutes: totalMinutes,
                muscleGroupsHit: muscleGroups,
                exerciseNames: allExercises
            )
        }
    }

    /// Detects trends from weekly summaries.
    ///
    /// - Parameter summaries: Weekly summaries (oldest-first).
    /// - Returns: Detected trends.
    public static func detectTrends(from summaries: [WeeklySummary]) -> [LoadTrend] {
        guard summaries.count >= 2 else { return [] }

        var trends: [LoadTrend] = []

        // Frequency analysis
        let counts = summaries.map(\.workoutCount)
        let recentAvg = Double(counts.suffix(2).reduce(0, +)) / 2.0
        let olderAvg = Double(counts.prefix(summaries.count - 2 < 1 ? 1 : summaries.count - 2).reduce(0, +))
            / Double(max(1, summaries.count - 2))

        if recentAvg < olderAvg * 0.6 && olderAvg > 0 {
            trends.append(LoadTrend(
                type: .frequencyDrop,
                message: "Your workout frequency dropped from ~\(Int(olderAvg.rounded())) to ~\(Int(recentAvg.rounded())) sessions/week.",
                severity: .warning
            ))
        } else if recentAvg > olderAvg * 1.3 && olderAvg > 0 {
            trends.append(LoadTrend(
                type: .frequencyRise,
                message: "Great consistency — you're training ~\(Int(recentAvg.rounded())) sessions/week, up from ~\(Int(olderAvg.rounded())).",
                severity: .positive
            ))
        }

        // Overreaching check (>6 sessions/week sustained for 2+ weeks)
        let highVolumeWeeks = counts.suffix(2).filter { $0 >= 6 }.count
        if highVolumeWeeks >= 2 {
            trends.append(LoadTrend(
                type: .overreaching,
                message: "You've trained 6+ times/week for 2 consecutive weeks. Consider a lighter week.",
                severity: .warning
            ))
        }

        // Undertraining check (<2 sessions/week for 2+ weeks)
        let lowVolumeWeeks = counts.suffix(2).filter { $0 <= 1 }.count
        if lowVolumeWeeks >= 2 {
            trends.append(LoadTrend(
                type: .undertraining,
                message: "Only 0-1 sessions/week recently. Even 2-3 short sessions help maintain progress.",
                severity: .warning
            ))
        }

        // Muscle group balance
        if let lastWeek = summaries.last {
            let majorGroups: Set<String> = ["Upper Push", "Upper Pull", "Lower", "Core"]
            let missing = majorGroups.subtracting(lastWeek.muscleGroupsHit)
            if !missing.isEmpty && lastWeek.workoutCount >= 3 {
                let missingList = missing.sorted().joined(separator: ", ")
                trends.append(LoadTrend(
                    type: .muscleImbalance,
                    message: "Last week missed: \(missingList). Try to hit all major groups each week.",
                    severity: .info
                ))
            }
        }

        // Push/pull ratio imbalance
        if let lastWeek = summaries.last, lastWeek.workoutCount >= 2 {
            let ratio = pushPullRatio(from: lastWeek)
            if ratio.push > 0 && ratio.pull > 0 {
                if ratio.push > ratio.pull * 2 {
                    trends.append(LoadTrend(
                        type: .pushPullImbalance,
                        message: "Push/Pull ratio is \(ratio.push):\(ratio.pull) this week. Aim for roughly 1:1 to prevent shoulder and posture issues.",
                        severity: .warning
                    ))
                } else if ratio.pull > ratio.push * 2 {
                    trends.append(LoadTrend(
                        type: .pushPullImbalance,
                        message: "Pull/Push ratio is \(ratio.pull):\(ratio.push) this week. Add some pressing to maintain balance.",
                        severity: .warning
                    ))
                }
            }
        }

        // Missed session pattern detection
        if summaries.count >= 4 {
            let pattern = detectMissedSessionPattern(counts: counts)
            if let pattern {
                trends.append(pattern)
            }
        }

        // Consistency
        if trends.isEmpty && summaries.count >= 3 {
            let variance = counts.variance()
            if variance < 1.5 {
                trends.append(LoadTrend(
                    type: .consistent,
                    message: "Solid consistency over the last \(summaries.count) weeks. Keep it up.",
                    severity: .positive
                ))
            }
        }

        return trends
    }

    // MARK: - Push/Pull Ratio

    /// Computes the count of Upper Push vs Upper Pull exercises in a weekly summary.
    public static func pushPullRatio(from summary: WeeklySummary) -> (push: Int, pull: Int) {
        var push = 0
        var pull = 0
        for name in summary.exerciseNames {
            let group = classifyMuscleGroup(name)
            if group == "Upper Push" { push += 1 }
            if group == "Upper Pull" { pull += 1 }
        }
        return (push, pull)
    }

    // MARK: - Missed Session Pattern Detection

    /// Looks for repeating drop-off patterns in workout counts.
    /// E.g., [4, 4, 1, 4, 4, 1] → drop-off every 3rd week.
    private static func detectMissedSessionPattern(counts: [Int]) -> LoadTrend? {
        guard counts.count >= 4 else { return nil }

        let avg = Double(counts.reduce(0, +)) / Double(counts.count)
        guard avg > 1 else { return nil }

        // Find weeks that are significantly below average (< 40% of avg)
        let dropThreshold = avg * 0.4
        let dropWeeks = counts.enumerated().filter { Double($0.element) < dropThreshold }

        // Need at least 2 drop weeks to detect a pattern
        guard dropWeeks.count >= 2 else { return nil }

        // Check if drops are evenly spaced
        let gaps = zip(dropWeeks, dropWeeks.dropFirst()).map { $1.offset - $0.offset }
        guard let firstGap = gaps.first, firstGap >= 2 else { return nil }
        let allSameGap = gaps.allSatisfy { $0 == firstGap }

        if allSameGap {
            let ordinal = firstGap == 2 ? "3rd" : "\(firstGap + 1)th"
            return LoadTrend(
                type: .missedSessionPattern,
                message: "You tend to drop off every \(ordinal) week. Consider scheduling a deliberate lighter week instead of skipping.",
                severity: .info
            )
        }

        return nil
    }

    // MARK: - Muscle Group Classification

    /// Maps an exercise name to a broad muscle group category.
    public static func classifyMuscleGroup(_ exerciseName: String) -> String? {
        let name = exerciseName.lowercased()

        // Lower body
        let lowerKeywords = ["squat", "deadlift", "lunge", "leg", "hip thrust",
                             "glute", "hamstring", "calf", "step-up", "box jump"]
        if lowerKeywords.contains(where: { name.contains($0) }) {
            return "Lower"
        }

        // Upper push
        let pushKeywords = ["bench", "press", "push-up", "pushup", "dip",
                            "overhead", "shoulder", "jerk", "thruster"]
        if pushKeywords.contains(where: { name.contains($0) }) {
            return "Upper Push"
        }

        // Upper pull
        let pullKeywords = ["row", "pull-up", "pullup", "chin-up", "lat",
                            "curl", "snatch", "clean", "face pull", "muscle-up"]
        if pullKeywords.contains(where: { name.contains($0) }) {
            return "Upper Pull"
        }

        // Core
        let coreKeywords = ["sit-up", "situp", "plank", "crunch", "ab",
                            "toes-to-bar", "l-sit", "hollow", "v-up"]
        if coreKeywords.contains(where: { name.contains($0) }) {
            return "Core"
        }

        return nil
    }
}

// MARK: - Array Extension

private extension Array where Element == Int {
    func variance() -> Double {
        guard count > 1 else { return 0 }
        let mean = Double(reduce(0, +)) / Double(count)
        let sumSquares = reduce(0.0) { $0 + pow(Double($1) - mean, 2) }
        return sumSquares / Double(count)
    }
}

File: android/core/src/main/java/com/sundeefundee/core/domain/intelligence/WeeklyLoadAnalyzer.kt
package com.sundeefundee.core.domain.intelligence

import com.sundeefundee.core.model.CompletedWorkoutRecord
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.temporal.WeekFields
import java.util.Locale
import kotlin.math.pow
import kotlin.math.round

/**
 * Analyzes completed workout history to surface volume, intensity,
 * frequency, and muscle-group balance trends over rolling weeks.
 *
 * Pure domain logic -- no framework dependencies.
 */
object WeeklyLoadAnalyzer {

    // MARK: - Types

    /** Summary of a single training week. */
    data class WeeklySummary(
        /** The Monday that starts this ISO week. */
        val weekStartDate: LocalDate,
        /** Total number of workouts completed. */
        val workoutCount: Int,
        /** Total training duration in minutes. */
        val totalMinutes: Int,
        /** Distinct muscle groups trained (from exercise names). */
        val muscleGroupsHit: Set<String>,
        /** Exercises performed this week. */
        val exerciseNames: List<String>
    )

    /** A detected trend across multiple weeks. */
    data class LoadTrend(
        val type: TrendType,
        val message: String,
        val severity: TrendSeverity
    )

    /** Types of detectable training trends. */
    enum class TrendType {
        /** Workout frequency is declining. */
        FREQUENCY_DROP,
        /** Workout frequency is increasing nicely. */
        FREQUENCY_RISE,
        /** Muscle group imbalance detected. */
        MUSCLE_IMBALANCE,
        /** Possible overreaching (too much volume). */
        OVERREACHING,
        /** Undertraining (very low volume). */
        UNDERTRAINING,
        /** Consistent training pattern. */
        CONSISTENT,
        /** Push/pull ratio imbalance within a week. */
        PUSH_PULL_IMBALANCE,
        /** Repeating pattern of missed/dropped weeks. */
        MISSED_SESSION_PATTERN,
    }

    /** Severity level for a detected trend. */
    enum class TrendSeverity {
        INFO,
        WARNING,
        POSITIVE
    }

    // MARK: - Analysis

    /**
     * Builds weekly summaries from completed workout records.
     *
     * @param workouts Completed workout records.
     * @param weekCount Number of recent weeks to analyze (default 4).
     * @return Weekly summaries sorted oldest-first.
     */
    fun weeklySummaries(
        workouts: List<CompletedWorkoutRecord>,
        weekCount: Int = 4
    ): List<WeeklySummary> {
        val weekFields = WeekFields.of(Locale.getDefault())

        // Group workouts by ISO week start (Monday)
        val weekBuckets = mutableMapOf<LocalDate, MutableList<CompletedWorkoutRecord>>()

        for (workout in workouts) {
            val workoutDate = parseDate(workout.date)
            val weekStart = workoutDate.with(weekFields.dayOfWeek(), 1)
            weekBuckets.getOrPut(weekStart) { mutableListOf() }.add(workout)
        }

        // Build summaries for the most recent N weeks
        val sortedWeeks = weekBuckets.keys.sorted()
        val recentWeeks = sortedWeeks.takeLast(weekCount)

        return recentWeeks.map { weekStart ->
            val records = weekBuckets[weekStart] ?: emptyList()
            val allExercises = records.flatMap { it.exerciseNames }
            val muscleGroups = allExercises.mapNotNull { classifyMuscleGroup(it) }.toSet()
            val totalMinutes = records.mapNotNull { it.duration }.sum()

            WeeklySummary(
                weekStartDate = weekStart,
                workoutCount = records.size,
                totalMinutes = totalMinutes,
                muscleGroupsHit = muscleGroups,
                exerciseNames = allExercises
            )
        }
    }

    /**
     * Detects trends from weekly summaries.
     *
     * @param summaries Weekly summaries (oldest-first).
     * @return Detected trends.
     */
    fun detectTrends(summaries: List<WeeklySummary>): List<LoadTrend> {
        if (summaries.size < 2) return emptyList()

        val trends = mutableListOf<LoadTrend>()

        // Frequency analysis
        val counts = summaries.map { it.workoutCount }
        val recentAvg = counts.takeLast(2).sum().toDouble() / 2.0
        val olderCount = maxOf(1, counts.size - 2)
        val olderAvg = counts.dropLast(2).sum().toDouble() / olderCount

        if (recentAvg < olderAvg * 0.6 && olderAvg > 0) {
            trends.add(
                LoadTrend(
                    type = TrendType.FREQUENCY_DROP,
                    message = "Your workout frequency dropped from ~${round(olderAvg).toInt()} to ~${round(recentAvg).toInt()} sessions/week.",
                    severity = TrendSeverity.WARNING
                )
            )
        } else if (recentAvg > olderAvg * 1.3 && olderAvg > 0) {
            trends.add(
                LoadTrend(
                    type = TrendType.FREQUENCY_RISE,
                    message = "Great consistency -- you're training ~${round(recentAvg).toInt()} sessions/week, up from ~${round(olderAvg).toInt()}.",
                    severity = TrendSeverity.POSITIVE
                )
            )
        }

        // Overreaching check (>6 sessions/week sustained for 2+ weeks)
        val highVolumeWeeks = counts.takeLast(2).count { it >= 6 }
        if (highVolumeWeeks >= 2) {
            trends.add(
                LoadTrend(
                    type = TrendType.OVERREACHING,
                    message = "You've trained 6+ times/week for 2 consecutive weeks. Consider a lighter week.",
                    severity = TrendSeverity.WARNING
                )
            )
        }

        // Undertraining check (<2 sessions/week for 2+ weeks)
        val lowVolumeWeeks = counts.takeLast(2).count { it <= 1 }
        if (lowVolumeWeeks >= 2) {
            trends.add(
                LoadTrend(
                    type = TrendType.UNDERTRAINING,
                    message = "Only 0-1 sessions/week recently. Even 2-3 short sessions help maintain progress.",
                    severity = TrendSeverity.WARNING
                )
            )
        }

        // Muscle group balance
        val lastWeek = summaries.lastOrNull()
        if (lastWeek != null) {
            val majorGroups = setOf("Upper Push", "Upper Pull", "Lower", "Core")
            val missing = majorGroups - lastWeek.muscleGroupsHit
            if (missing.isNotEmpty() && lastWeek.workoutCount >= 3) {
                val missingList = missing.sorted().joinToString(", ")
                trends.add(
                    LoadTrend(
                        type = TrendType.MUSCLE_IMBALANCE,
                        message = "Last week missed: $missingList. Try to hit all major groups each week.",
                        severity = TrendSeverity.INFO
                    )
                )
            }
        }

        // Push/pull ratio imbalance
        if (lastWeek != null && lastWeek.workoutCount >= 2) {
            val ratio = pushPullRatio(lastWeek)
            if (ratio.first > 0 && ratio.second > 0) {
                if (ratio.first > ratio.second * 2) {
                    trends.add(
                        LoadTrend(
                            type = TrendType.PUSH_PULL_IMBALANCE,
                            message = "Push/Pull ratio is ${ratio.first}:${ratio.second} this week. Aim for roughly 1:1 to prevent shoulder and posture issues.",
                            severity = TrendSeverity.WARNING
                        )
                    )
                } else if (ratio.second > ratio.first * 2) {
                    trends.add(
                        LoadTrend(
                            type = TrendType.PUSH_PULL_IMBALANCE,
                            message = "Pull/Push ratio is ${ratio.second}:${ratio.first} this week. Add some pressing to maintain balance.",
                            severity = TrendSeverity.WARNING
                        )
                    )
                }
            }
        }

        // Missed session pattern detection
        if (summaries.size >= 4) {
            detectMissedSessionPattern(counts)?.let { trends.add(it) }
        }

        // Consistency
        if (trends.isEmpty() && summaries.size >= 3) {
            val variance = counts.variance()
            if (variance < 1.5) {
                trends.add(
                    LoadTrend(
                        type = TrendType.CONSISTENT,
                        message = "Solid consistency over the last ${summaries.size} weeks. Keep it up.",
                        severity = TrendSeverity.POSITIVE
                    )
                )
            }
        }

        return trends
    }

    // MARK: - Push/Pull Ratio

    /** Computes the count of Upper Push vs Upper Pull exercises in a weekly summary. */
    fun pushPullRatio(summary: WeeklySummary): Pair<Int, Int> {
        var push = 0
        var pull = 0
        for (name in summary.exerciseNames) {
            when (classifyMuscleGroup(name)) {
                "Upper Push" -> push++
                "Upper Pull" -> pull++
            }
        }
        return push to pull
    }

    // MARK: - Missed Session Pattern Detection

    /**
     * Looks for repeating drop-off patterns in workout counts.
     * E.g., [4, 4, 1, 4, 4, 1] -> drop-off every 3rd week.
     */
    private fun detectMissedSessionPattern(counts: List<Int>): LoadTrend? {
        if (counts.size < 4) return null

        val avg = counts.sum().toDouble() / counts.size
        if (avg <= 1) return null

        // Find weeks that are significantly below average (< 40% of avg)
        val dropThreshold = avg * 0.4
        val dropWeeks = counts.mapIndexedNotNull { index, value ->
            if (value.toDouble() < dropThreshold) index else null
        }

        // Need at least 2 drop weeks to detect a pattern
        if (dropWeeks.size < 2) return null

        // Check if drops are evenly spaced
        val gaps = dropWeeks.zipWithNext { a, b -> b - a }
        val firstGap = gaps.firstOrNull() ?: return null
        if (firstGap < 2) return null
        val allSameGap = gaps.all { it == firstGap }

        if (allSameGap) {
            val ordinal = if (firstGap == 2) "3rd" else "${firstGap + 1}th"
            return LoadTrend(
                type = TrendType.MISSED_SESSION_PATTERN,
                message = "You tend to drop off every $ordinal week. Consider scheduling a deliberate lighter week instead of skipping.",
                severity = TrendSeverity.INFO
            )
        }

        return null
    }

    // MARK: - Muscle Group Classification

    /** Maps an exercise name to a broad muscle group category. */
    fun classifyMuscleGroup(exerciseName: String): String? {
        val name = exerciseName.lowercase()

        // Lower body
        val lowerKeywords = listOf(
            "squat", "deadlift", "lunge", "leg", "hip thrust",
            "glute", "hamstring", "calf", "step-up", "box jump"
        )
        if (lowerKeywords.any { name.contains(it) }) {
            return "Lower"
        }

        // Upper push
        val pushKeywords = listOf(
            "bench", "press", "push-up", "pushup", "dip",
            "overhead", "shoulder", "jerk", "thruster"
        )
        if (pushKeywords.any { name.contains(it) }) {
            return "Upper Push"
        }

        // Upper pull
        val pullKeywords = listOf(
            "row", "pull-up", "pullup", "chin-up", "lat",
            "curl", "snatch", "clean", "face pull", "muscle-up"
        )
        if (pullKeywords.any { name.contains(it) }) {
            return "Upper Pull"
        }

        // Core
        val coreKeywords = listOf(
            "sit-up", "situp", "plank", "crunch", "ab",
            "toes-to-bar", "l-sit", "hollow", "v-up"
        )
        if (coreKeywords.any { name.contains(it) }) {
            return "Core"
        }

        return null
    }

    // MARK: - Private Helpers

    /** Parse an ISO8601 date string to LocalDate. Tolerant of multiple ISO formats. */
    private fun parseDate(dateString: String): LocalDate {
        return try {
            LocalDate.parse(dateString.substring(0, 10))
        } catch (_: Exception) {
            try {
                LocalDateTime.parse(dateString).toLocalDate()
            } catch (_: Exception) {
                LocalDate.now()
            }
        }
    }

    /** Compute population variance of a list of integers. */
    private fun List<Int>.variance(): Double {
        if (size <= 1) return 0.0
        val mean = sum().toDouble() / size
        return sumOf { (it.toDouble() - mean).pow(2) } / size
    }
}
